File: University Coursework/Extracted Submissions/SENG2200_PA2/src/main/java/LinkedList.java
import java.util.*;

/**
 * Circular Doubly Linked List with a Single Sentinel
 * Contains methods to add/remove at relevant indexes and query methods
 * Contains private classes such as DataNode<T>, SentinelNode<T> and Iterator<T>
 */
public class LinkedList<T> implements Iterable<T>
{
    private Node<T> sentinel;
    private int numberOfNodes;
    private int modCount;

    /**
     * public LinkedList()
     * Default Constructor
     * Sets numberOfNodes to Zero and creates the links for the sentinel node to point to itself
     */
    public LinkedList()
    {
        numberOfNodes = 0;
        modCount = 0;

        sentinel = new SentinelNode<>();
        sentinel.setNext(sentinel);
        sentinel.setPrevious(sentinel);
    }

    /**
     * public void prepend(T data) //Add to Head method
     * Calls the addBefore method, adding after the sentinel node.getNext()
     * @param data - Generic Data Type
     */
    public void prepend(T data) //Add to Head
    {
        addBefore(data,sentinel.getNext());
    }

    /**
     * public void append(T data) //Add to Tail method
     * Calls the  addBefore method, adding before the sentinel node
     * @param data - Generic Data type
     */
    public void append(T data) //Add to tail
    {
        addBefore(data, sentinel);
    }

    /**
     * private void addBefore(T data, Node<T> currentNode)
     * Creates a new node with the parameter data before the specified current Node
     * Updates the appropriate links to accommodate the new node
     * @param data - Generic Data Type
     * @param currentNode - Current Node object to insert before
     */
    private void addBefore(T data, Node<T> currentNode)
    {
        Node<T> previousNode = currentNode.getPrevious();
        Node<T> newNode = new DataNode<>(data, currentNode, previousNode);

        previousNode.setNext(newNode);
        currentNode.setPrevious(newNode);

        numberOfNodes++;
        modCount++;
    }

    /**
     * public void insert(T data, int index) //AddAt method
     * Creates a new node with the parameter data at one position before the specified index parameter
     * If the index is position 0, it will prepend the data.
     * If the index is equal to the size of the LL, it will append the data
     * Updates the appropriate links to accommodate the new node
     * @param data - Generic Data type
     * @param index - Integer specifying the index in the LL
     */
    public void insert(T data, int index)
    {
        if(index == 0)
            prepend(data); //Without this check it would attempt to place the node at position -1 (Doesn't exist)

        else if(index == size())
            append(data);

        else
        {
            int counter = 0;
            checkBounds(index);

            //If in range, continue to find the index
            for (Node<T> cursor = sentinel.getNext(); cursor != sentinel; cursor = cursor.getNext())
            {
                //If the counter has matched the index
                if (counter == index)
                {
                    addBefore(data, cursor);
                    return;
                }
                //If not, increment and continue to loop
                else
                    counter++;
            }
        }
    }

    /**
     * public T removeFromHead()
     * Calls the removeNode method to remove sentinel.getNext()
     * @return - T data of the deleted node
     */
    public T removeFromHead()
    {
        return removeNode(sentinel.getNext());
    }

    /**
     * public T removeFromTail()
     * Calls the removeNode method to remove sentinel.getPrevious()
     * @return - T data of the deleted node
     */
    public T removeFromTail()
    {
        return removeNode(sentinel.getPrevious());
    }

    /**
     * private T removeNode(Node<T> deleteNode)
     * Checks if the list is empty first - throws error if no nodes exist in the list
     * If not empty - method will specified node
     * Updates links to accommodate for the removal of a node
     * @param deleteNode - Node to be deleted
     * @return
     */
    private T removeNode(Node<T> deleteNode)
    {
        if(isEmpty())
            throw new NoSuchElementException("Error. No Nodes in the List");

        T nodeData = deleteNode.getData();
        Node<T> nextNode = deleteNode.getNext();
        Node<T> previousNode = deleteNode.getPrevious();

        previousNode.setNext(nextNode);
        nextNode.setPrevious(previousNode);

        // Clean up delete node references
        deleteNode.setData(null);

        // If iterator cursor is pointing to deleteNode and we concurrently set currentNode/previous to null,
        // the iterator will break
        deleteNode.setNext(null);
        deleteNode.setPrevious(null);

        numberOfNodes--;
        modCount++;
        return nodeData;
    }

    /**
     * public int indexOf(T data)
     * Searches the linked list using a Node<T> cursor and an integer counter to match the cursor's data with the input T data
     * If cases are equal it will return the integer counter
     * If not found in the LL, it will return -1
     * @param data - Generic Type T
     * @return - Integer of index where the data was found (-1 if not in LL)
     */
    public int indexOf(T data)
    {
        int counter = 0;
        for(Node<T> cursor = sentinel.getNext(); cursor != sentinel; cursor = cursor.getNext())
        {
            if (data == null && cursor.getData() == null)
                return counter;
            else if(cursor.getData().equals(data))
                return counter; //Return position in the list
            else
                counter++; //Increment until found
        }
        return -1; //If not found in the list
    }

    /**
     * public T get(int index)
     * Searches the linked list using a Node<T> cursor and an integer counter to find iterate up to the specified index
     * Once the index has been reached, it will return the Node<T> cursor's data
     * @param index - Integer specifying what position the data will be returned from
     * @return - T data that can be type cast to any data type
     */
    public T get(int index)
    {
        checkBounds(index);

        int counter = 0;
        for(Node<T> cursor = sentinel.getNext(); cursor != sentinel; cursor = cursor.getNext())
        {
            if(counter == index)
                return cursor.getData();
            else
                counter++;
        }
        return null; //Shouldn't return null due to boundary check before the for loop
    }

    /**
     * private void checkBounds(int index)
     * Checks if the parameter index is out of range - If out of range, throws a new IndexOutOfBoundsException
     * @param index - Index of node
     */
    private void checkBounds(int index)
    {
        if(index < 0 || index >= size())
            throw new IndexOutOfBoundsException(String.format("Error. Index %d is out of bounds [0,%d)",index, size()));
    }


    /**
     * public void clear()
     * Removes all nodes from the linked list whilst it is not empty
     */
    public void clear()
    {
        while(!isEmpty())
            removeFromHead();
    }

    /**
     * public boolean isEmpty()
     * Checks if there are no nodes in the linked list
     * @return - Boolean value
     */
    public boolean isEmpty()
    {
        return numberOfNodes == 0;
    }

    /**
     * public int size()
     * Returns the number of nodes in the linked list
     * @return - Integer value
     */
    public int size()
    {
        return numberOfNodes;
    }

    /**
     * @Override
     * public String toString()
     * Overridden toString method to accommodate for the assignment specification
     * Iterates through the linked list of polygons and appends to the end of a StringBuilder
     * @return - String of all polygons found in the LL
     */
    @Override
    public String toString()
    {
        StringBuilder sb = new StringBuilder();
        for(Node<T> cursor = sentinel.getNext(); cursor != sentinel; cursor = cursor.getNext())
        {
            sb.append(cursor.getData())
              .append("\n");
        }

        return sb.toString();
    }

    /**
     * Constructor for Iterator
     * Creates new Iterator object
     * @return
     */
    public Iterator<T> iterator(){return new MyIterator<>();}

    /**
     * PRIVATE CLASS MYITERATOR
     *  Generic Iterator class that only exists inside Linked List Class
     */
    private class MyIterator<E> implements Iterator<T>
    {
        private Node<T> currentNode;
        private int expectedModCount;

        /**
         * Default constructor
         */
        private MyIterator()
        {
            currentNode = sentinel;
            expectedModCount = modCount;
        }

        /**
         * public boolean hasNext()
         * Overridden method from Iterator Interface
         * Determines if there is another node following the current node
         * @return - Boolean if node exists
         */
        @Override
        public boolean hasNext(){ return (currentNode.getNext() != sentinel); }

        /**
         * public T next()
         * Overridden method from Iterator Interface
         * Moves to next node
         * @return - Next Node's Data
         */
        @Override
        public T next()
        {
            checkForConcurrentModification();
            if(!hasNext())
                throw new NoSuchElementException("Element does not exist");
            currentNode = currentNode.getNext();
            return currentNode.getData();
        }

        /**
         * public void remove()
         * Overridden method from Iterator Interface
         * Removes the current node
         */
        @Override
        public void remove()
        {
            /*
            From http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/Iterator.java#Iterator.remove%28%29

            Removes from the list the last element that was returned by next() or previous() (optional operation).
            This call can only be made once per call to next or previous. It can be made only if
            add(java.lang.Object) has not been called after the last call to next or previous.
            */

            checkForConcurrentModification();
            Node<T> previousNode = currentNode.getPrevious();

            LinkedList.this.removeNode(currentNode);

            // to preserve the linked list order we have to move the iterator back
            //Reassign current node to previous node
            currentNode = previousNode;

            //Linkages are handled by the removeNode method
            // modCount will be updated in LinkedList.remove()
            // have to update expectedModCount to allow continued use of iterator
            expectedModCount++;
        }

        /**
         * private void checkForConcurrentModification()
         * Determines whether the mod count is exact to the expected mod count
         */
        private void checkForConcurrentModification()
        {
            if(modCount != expectedModCount)
                throw new ConcurrentModificationException("Concurrent modification on Linked List is not permissible.");
        }

    }




    /**
     * PRIVATE CLASS DATANODE<T>
     *     Generic Private DataNode class that only exists inside the Linked List Class
     */
    private class DataNode<T> implements Node<T>
    {
        private T _data;
        private Node<T> _next;
        private Node<T> _previous;

        /**
         * public DataNode()
         * Default Constructor
         * Sets all private data to null
         */
        public DataNode()
        {
            _data = null;
            _next =  null;
            _previous = null;
        }

        /**
         * public DataNode(T data, Node<T> next, Node<T> previous)
         * Overloaded Constructor
         * @param data - Type T that can be type cast to any data type
         * @param next - Node<T> T defining the currentNode node to follow in the LL
         * @param previous - Node<T> T defining the previous node to follow in the LL
         */
        public DataNode(T data, Node<T> next, Node<T> previous)
        {
            this._data = data;
            this._next = next;
            this._previous = previous;
        }

        /**
         * public T getData()
         * @return - T Data of private _data variable
         */
        public T getData() {return _data;}

        /**
         * public Node<T> getNext()
         * @return - Node<T> T of private _next variable
         */
        public Node<T> getNext() {return _next;}

        /**
         * public Node<T> getPrevious()
         * @return - Node<T> T of private _previous variable
         */
        public Node<T> getPrevious() {return _previous;}

        /**
         * public void setData(T data)
         * @param data - Type T that set the private _data variable to the parameter
         */
        public void setData(T data) {this._data = data;}

        /**
         * public void setNext(Node<T> currentNode)
         * @param next - Type Node<T> that sets the private _next variable to the parameter
         */
        public void setNext(Node<T> next) {this._next = next;}

        /**
         * public void setPrevious(Node<T> previous)
         * @param previous - Type Node<T> that sets the private _previous variable to the parameter
         */
        public void setPrevious(Node<T> previous) {this._previous = previous;}
    }

    /**
     * PRIVATE CLASS SENTINELNODE<T>
     *     Generic Private SentinelNode class that only exists inside the Linked List Class
     */
    private class SentinelNode<T> implements Node<T>
    {
        private Node<T> _next;
        private Node<T> _previous;

        /**
         * public SentinelNode()
         * Default Constructor
         * Sets all private data to null
         */
        public SentinelNode()
        {
            _next =  null;
            _previous = null;
        }

        /**
         * public SentinelNode(Node<T> currentNode, Node<T> previous)
         * Overloaded Constructor
         * @param next - Node<T> T defining the currentNode node to follow in the LL
         * @param previous - Node<T> T defining the previous node to follow in the LL
         */
        public SentinelNode(Node<T> next, Node<T> previous)
        {
            this._next = next;
            this._previous = previous;
        }

        /**
         * public T getData()
         * @return - Throw exception since sentinel node should have no data
         */
        public T getData() {throw new UnsupportedOperationException("Trying to access sentinel node");}

        /**
         * public Node<T> getNext()
         * @return - Node<T> T of private _next variable
         */
        public Node<T> getNext() {return _next;}

        /**
         * public Node<T> getPrevious()
         * @return - Node<T> T of private _previous variable
         */
        public Node<T> getPrevious() {return _previous;}

        /**
         * public void setData(T data)
         * Throw exception since sentinel node should not set data
         * @param data - Type T that set the private _data variable to the parameter
         */
        public void setData(T data) {throw new UnsupportedOperationException("Trying to set data on sentinel");}

        /**
         * public void setNext(Node<T> currentNode)
         * @param next - Type Node<T> that sets the private _next variable to the parameter
         */
        public void setNext(Node<T> next) {this._next = next;}

        /**
         * public void setPrevious(Node<T> previous)
         * @param previous - Type Node<T> that sets the private _previous variable to the parameter
         */
        public void setPrevious(Node<T> previous) {this._previous = previous;}
    }

}

File: University Coursework/Extracted Submissions/SENG1120_PA1/Seng1120Assignment1/LinkedList.cpp
#include "LinkedList.h"

/*
 * Author: Jordan Haigh		StNo: 3256730		Lab: Thurs 11am
 * Course: SENG1120
 * LinkedList.cpp
 */

namespace SENG1120ASSIGN
{
    LinkedList::LinkedList()
    {
        //I have set up my linked list differently to the lecture slides
		//My head and tail pointers will be sentinels - they will always be NULL
		//All data will go between the head and the tail node

		//Initialise head and tail
		head = new Node();
        tail = new Node();

        //Link head and tail together
        head->setNext(tail);
        tail->setPrevious(head);

        //Initialise counter - Makes it easier to find size of linked list without traversing
        listLength = 0;
    }

    LinkedList::~LinkedList()
    {
        //Call clear to remove the innards
        clear();

		//No more data inside the head and the tail. Set the linkages for the head and tail nodes
		head->setNext(NULL);
		tail->setPrevious(NULL);

        //Delete head and tail to remove the sentinel nodes
        delete head;
        delete tail;
    }

    bool LinkedList::isEmpty()
    {
        //If list length is less than 1 (Nothing between two sentinels), will return false
        return listLength<1;
    }

    void LinkedList::addAt(std::size_t index, valueType data)
    {
        //Assuming that user is inputting a new node in between two existing nodes
        //Adds node at before the specified index so that the new node becomes that index

        //Traverse to the index
		//Uses a counter variable to check whether the cursor is at the index or not
		//Double check first that the index entered by the user is within range
        std::size_t counter = 0;
        if(index < 0 || index >=size())
        {
            throw "ERROR. Index is out of range";
        }
		//If in range, continue to find the index
        for(Node* cursor = head->getNext(); cursor != tail; cursor = cursor->getNext())
        {
            //If the counter has matched the index
			if(counter == index)
            {
                //Create new node
                Node* previousNode = cursor->getPrevious();
                //New Node
                Node* newNode = new Node(data, cursor, previousNode);
                //Update links
                cursor->setPrevious(newNode);
                previousNode->setNext(newNode);
                listLength++;
            }
			//If not, increment and continue to loop
			else
				counter++;
        }
    }

    LinkedList::valueType LinkedList::removeAt(std::size_t index)
    {
		//Uses a counter variable to check whether the cursor is at the index or not
		//Double check first that the index entered by the user is within range
		std::size_t counter = 0;
        if(index <0 || index >= size())
        {
            throw "ERROR: Index is out of range.";
        }

		//If in range, continue to find the index
        for(Node* cursor = head->getNext(); cursor != tail; cursor = cursor->getNext())
        {
            //If counter matches with the index
			if(counter == index)
            {
                //Found index of Node
                //Create a previousNode and nextNode to fix the links when removing the specified Node
                Node* previousNode = cursor->getPrevious();
                Node* nextNode = cursor->getNext();

                previousNode->setNext(nextNode);
                nextNode->setPrevious(previousNode);
                valueType cursorData = cursor->getData();

				//Remove the links of the current cursor to avoid a stack overflow
				cursor->setNext(NULL);
				cursor->setPrevious(NULL);
				//Delete the current cursor
                delete cursor;

                // Decrement size
                listLength--;

                return cursorData;
            }
            //If not, increment the counter and continue to loop
			else
            {
                counter++;
            }
        }

		//Shouldn't be able to reach this section due to the boundaries check
        return NULL;
    }

    bool LinkedList::removeFirst(valueType data)
    {

        // Removes first instance of data
        for(Node* cursor = head->getNext(); cursor != tail; cursor = cursor->getNext())
        {
            //Iterate until found data

            //if(data == NULL ? cursor->getData() == NULL : cursor->getData().equals(data))
            //This ternary could have worked, though because the typedef is not a pointer, this isn't achievable
            //TODO replace with object.equals at a later date

			//Uses compare method to compare the two data types (Similar to .equals in Java)
            if(cursor->getData().compare(data) == 0)
            {
                //If strings are equals, will result to 0 (Boolean)
                //Create nextNode and previousNode and update linkages
                Node* nextNode = cursor->getNext();
                Node* previousNode = cursor->getPrevious();

                previousNode->setNext(nextNode);
                nextNode->setPrevious(previousNode);

				//Remove the pointers of the cursor to avoid a stack overflow
				cursor->setNext(NULL);
				cursor->setPrevious(NULL);
				//Remove cursor from the linked list
                delete cursor;

                // Decrement size
                listLength--;
				return true;
            }
        }
		return false;
    }

    LinkedList::valueType LinkedList::get(std::size_t index)
    {
		//Uses a counter variable to check whether the cursor is at the index or not
		//Double check first that the index entered by the user is within range
		std::size_t counter=0;
        if(index < 0 || index >= size())
        {
            throw "ERROR. Index is out of range";
        }
		//If in range, start to loop until found correct index
        for(Node* cursor = head->getNext(); cursor != tail; cursor = cursor->getNext())
        {
            //If counter is equal to the index
			if(counter == index)
            {
                return cursor->getData();
            }
            //If not, increment and continue to loop
			else
                counter++;
        }
        //should never be returned due to bounds check
        return NULL;
    }

    void LinkedList::set(std::size_t index, valueType data)
    {
		//Uses a counter variable to check whether the cursor is at the index or not
		//Double check first that the index entered by the user is within range
		std::size_t counter = 0;
        if(index < 0 || index >= size())
        {
            throw "ERROR. Index in out of range";
        }
        //If in range, start to loop until found correct index
		for(Node* cursor = head->getNext(); cursor != tail; cursor = cursor->getNext())
        {
            //If found correct index
			if(counter == index)
            {
                //Set new cursor data
				cursor->setData(data);
				return;
            }
            //If not found, increment and continue to loop
			else
                counter++;
        }
    }

    std::size_t LinkedList::indexOf(valueType data)
    {
		//Start counter at position 0
		std::size_t counter = 0;
		//Start loop of the linked list
        for(Node* cursor = head->getNext(); cursor != tail; cursor = cursor->getNext())
        {
            //Data has to be exactly the same (including capitals)
            //Need to use .compare to compare two strings as CPP doesn't include a .equals() alternative
            //TODO replace with object.equals at a later date
            
			//If data matches with the argument data
			if(cursor->getData().compare(data) == 0)
            {
                return counter;
            }
            //Increment counter until found
			else
                counter++;
        }
		//Return -1 if not found in the linked list
        return -1;
    }

    void LinkedList::addToHead(valueType& data)
    {
		//Create newNode that will be the next node after head
		Node* nextNode = head->getNext();
		//Use Node constructor to initialise data and links
        Node* newNode = new Node(data, nextNode, head);

        //Update links
        head->setNext(newNode);
        nextNode->setPrevious(newNode);
		//Increment list length
        listLength++;
    }

    void LinkedList::addToTail(valueType& data)
    {
        //If the linked list is empty, the previous node will be head
		//If the linked list is not empty, the previous node will be tail->getPrevious()
        Node* previousNode = tail->getPrevious();

        //Create new node with data and links
        Node* newNode = new Node(data,tail,previousNode);

        //Update links
        tail->setPrevious(newNode);
        previousNode->setNext(newNode);
		//Increment list length
        listLength++;
    }

    void LinkedList::removeFromHead()
    {
        //If the linked list is empty (No nodes between head and tail), just return
		if(listLength == 0)
            return;
        //If not, continue with the removal
		else
        {
            //Create two nodes for the removal: deleteNode (Node that will be removed)
			//									nextNode (The next node after the node to be deleted)
			Node* deleteNode = head->getNext();
            Node* nextNode = deleteNode->getNext();

            //Update links
            head->setNext(nextNode);
            nextNode->setPrevious(head);

			//Remove the links from deleteNode to avoid stack overflow
			deleteNode->setNext(NULL);
			deleteNode->setPrevious(NULL);
            // Delete node
            delete deleteNode; //Java API differs from this as it is meant to return data

            // Decrement size
            listLength--;
        }
    }

    void LinkedList::removeFromTail()
    {
        //If the linked list is empty (No nodes between head and tail), just return
		if(listLength == 0)
            return;
        //If not empty, continue with the method
		else
        {
			//Create two nodes for the removal: deleteNode (Node that will be removed)
			//									previousNode (The previous node before the node to be deleted)
			Node* deleteNode = tail->getPrevious();
            Node* previousNode = deleteNode->getPrevious();

            //Update links
            tail->setPrevious(previousNode);
            previousNode->setNext(tail);

			//Remove the links from the deleted node to avoid stack overflow
			deleteNode->setNext(NULL);
			deleteNode->setPrevious(NULL);
            // Delete node
            delete deleteNode;

            // Decrement size
            listLength--;
        }
    }

    std::size_t LinkedList::size()
    {
        //Return the current list length
		return listLength;
    }

    void LinkedList::clear()
    {
        //Clears the innards of the linked list without removing the head and tail sentinel nodes
        while(!isEmpty())
            removeFromHead();
    }

	LinkedList::valueType LinkedList::toString() const
	{
		valueType output;
		//Overloaded cout operator to print the node data in the linked list
		for (Node* cursor = head->getNext(); cursor != tail; cursor = cursor->getNext())
		{
			//iterate over all nodes and append to output stream
			output.append(cursor->getData());
			output.append(" ");
		}
		return output;
	}

    //Iterator methods
	//Methods not currently used in the program. May be implemented at a later date
    void LinkedList::iteratorReset()
    {
        //Reset the cursor to the first node between head and tail
		iteratorCursor = head->getNext();
    }

    void LinkedList::iteratorNext()
    {
        //Check whether there is another pointer beyond the current cursor
		//If so, move to the next Node in the linked list
		if(iteratorHasNext())
            iteratorCursor = iteratorCursor->getNext();
        //If not, throw an error
		else
            throw"ERROR. Iterator does not have next";
    }

    LinkedList::valueType LinkedList::iteratorData()
    {
        //Return the data of the current iterator
		return iteratorCursor->getData();
    }

    bool LinkedList::iteratorHasNext()
    {
        //Boolean result whether there is a next node that is not tail
		//If true, return true
		if(iteratorCursor->getNext()!=tail)
            return true;
        //Return false if not
		else
            return false;
    }

    std::ostream& operator << (std::ostream& output, const LinkedList& target)
    {

		return output << target.toString() << endl;
    }


}
